community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java

```diff
@@ -387,53 +418,87 @@ public Object getProperty( String key, Object defaultValue )
             return Collections.emptyMap();
         }
 
-        try ( Statement statement = actions.statement() )
+        try ( Statement ignore = actions.statement() )
         {
-            try ( Cursor<NodeItem> node = statement.readOperations().nodeCursorById( nodeId ) )
+            KernelTransaction transaction = actions.kernelTransaction();
+            CursorFactory cursors = transaction.cursors();
+            int itemsToReturn = keys.length;
+            Map<String,Object> properties = new HashMap<>( itemsToReturn );
+            Token token = transaction.token();
+
+            //Find ids, note we are betting on that the number of keys
+            //is small enough not to use a set here.
+            int[] propertyIds = new int[itemsToReturn];
+            for ( int i = 0; i < itemsToReturn; i++ )
             {
-                try ( Cursor<PropertyItem> propertyCursor = statement.readOperations().nodeGetProperties( node.get() ) )
+                String key = keys[i];
+                if ( key == null )
                 {
-                    return PropertyContainerProxyHelper.getProperties( statement, propertyCursor, keys );
+                    throw new NullPointerException( String.format( "Key %d was null", i + 1 ) );
                 }
+                propertyIds[i] = token.propertyKey( key );
```

Comment:
Perhaps my comment from earlier got lost in this back-and-forth, but I suspect it might still be profitable to sort the propertyKeyIds and use binary search for the contains check. The performance difference for small N should be negligible, while scaling much better as N grows large.

Reply:
No I remember, I think last time you commented was on labels but I guess same apply here. Have added a card for it. Let's add one thing at a time so it will be easier to track what is providing improvement and not.